### ISSUE TITLE
[Question] dialog does not appear, after click form-button-cancel (only works mode-debug)

### ISSUE THREAD

**Issue by AngelLopezBellmont** [microsoft/playwright#14118]
**Test works only mode-debug but not headed, Why?**

After click a form-button-cancel it should appears an alert (_Do you really want to discard the results?_). 
The test try to waitFor the alert **But it only works in mode-debug**. 
In the other modes, the whole form is closed without showing the alert. Why?

- playwright version: ^1.21.1
- operating system: locally-> windows, cicd -> azure
- auth tool: keycloak

```javascript
const {test, expect} = require('@playwright/test');
const myProject = 'demo-project-nuremberg';

test('click form-button-cancel does not show alert ', async ({page}) => {
  //login:
  await page.goto('https://dmitrygozman.fieldcode.com');
  await page.locator('[id="username"]').fill('dgozman.demo@gmail.com');
  await page.locator('[id="password"]').fill('Dgozman123');
  await page.locator('[id="fc-login-button"]').click();

  //open sidebar:
  await page.locator('svg[data-icon="angle-double-right"] >> nth=0').waitFor();
  await page.locator('svg[data-icon="angle-double-right"] >> nth=0').click();

  //click button-create-ticket:
  await page.locator('[id="sidebarMenu"] >> [title="Create ticket"]').waitFor();
  await page.locator('[id="sidebarMenu"] >> [title="Create ticket"]').click();

  //chose project to open form:
  const selector_myproject = '[data-ui-test="ticketCreationProject"] >> text='+myProject+'';
  await page.locator(selector_myproject).waitFor({state:"visible"});
  //await page.locator(selector_myproject).click({delay:1000, force:true});
  await page.locator(selector_myproject).click({delay:1000});

  //check form is open:
  await page.locator('text=Create ticket for project').waitFor({state:"visible"});
  await expect(page.locator('text=Create ticket for project')).toBeVisible;


  //click form-button-cancel:
  /*
  Here is the problem , it works in mode-debug but not --headed ,etc. Why?
  after click ->  await page.locator(buttonCancel).click({delay:500});
  the whole form disappears, without showing the alert-discard-results
  */
  const buttonCancel = 'button:has-text("Cancel")';
  await page.locator(buttonCancel).waitFor({state:"visible"});
  await page.locator(buttonCancel).focus();
  await page.locator(buttonCancel).hover({force:true});
  await page.locator(buttonCancel).click({delay:500});  // What am I doing wrong in this click????


  //ERROR HERE! playwright can't find the dialog locator 
  await page.locator('text=Do you really want to discard the results?').waitFor();
  await page.locator('button:has-text("Yes")').waitFor();
  await page.locator('button:has-text("Yes")').click({delay:500});

  //check form is closed:
  const selector = 'text=Create ticket for project';
  await page.locator(selector).waitFor({state:"detached"});
  await expect(page.locator(selector)).toHaveCount(0);

});


```

**Comment by ltsuda** [comment]
Please, take a look at these documentation links:

https://playwright.dev/docs/dialogs#api-reference
https://playwright.dev/docs/api/class-dialog

If the alert is of type `alert`, you will need to do this (from the links above):

```javascript
page.on('dialog', async dialog => {
    console.log(dialog.message());
    await dialog.dismiss(); // you can change the action you need, like `.accept({ promptText: "Yes" }) or `.dismiss()` for cancel actions
  });
```

**Comment by AngelLopezBellmont** [comment]
@ltsuda thanks for the info. But some questions:

- why is working in mode-debug. The behavior is different as headed?? 
- I had test with all code (without locators) and they were running ok opening the dialog etc How can it be?   umm unless there have been any changes in the UI with new React features)

**Comment by ltsuda** [comment]
Anytime.

Let's wait for someone from Playwright's team to answer your questions as I only used the framework for studying and never used the alert API.

**Comment by mxschmitt** [comment]
@ltsuda with dialog he means HTML dialogs, not normal alert/confirm dialogs.

---

@AngelLopezBellmont I was able to repro it, but was not yet able to get a minimal reproducible out of it yet.

**Comment by NagaAppariGlobal** [comment]
I have upgraded from v1.17 to  v 21.1 , i am seeing the same issue. pop up Dialog is not appearing after clicking on the HTML button . i have tried on the lower version till v1.20 .And it works till V1.20. Seeing this issue on V1.21 and V1.22

![image](https://user-images.githubusercontent.com/93261267/168326064-bb4cf1c1-89ae-423f-9d93-e6304471e6d4.png)

**Comment by mxschmitt** [comment]
@NagaAppariGlobal your issue is different. Please file a different issue, thanks.

**Comment by mxschmitt** [comment]
@AngelLopezBellmont for me it looks more like a race condition inside the application. Since if you put a ` await page.waitForTimeout(500);`before `const buttonCancel = 'button:has-text("Cancel")';` it works for me.

**Comment by AngelLopezBellmont** [comment]
@mxschmitt super yes you got it :) . With your new line, works to me as well.
```javascript
....
  //click form-button-cancel:
  /*
  Here is the problem , it works in mode-debug but not --headed ,etc. Why?
  after click ->  await page.locator(buttonCancel).click({delay:500});
  the whole form disappears, without showing the alert-discard-results
  */
  await page.waitForTimeout(500);                 //insert by mxschmitt
  const buttonCancel = 'button:has-text("Cancel")';
  await page.locator(buttonCancel).waitFor({state:"visible"});
  await page.locator(buttonCancel).focus();
  await page.locator(buttonCancel).hover({force:true});
```

To be honest I forgot to add in the issue-descrption, that in other branches is working with the old code without locators:

```javascript
  async close_formTicketCreation() {
    await this.page.waitForSelector('button:has-text("Cancel")');
    await this.page.click('button:has-text("Cancel")');
    await this.page.waitForSelector('text=Do you really want to discard the results?');
    await this.page.waitForSelector('button:has-text("Yes")');
    await this.page.click('button:has-text("Yes")');
  }
```
 
So yes @mxschmitt definnitely looks like it's a race condition. 
But now,  can we assume that the new feature locator  is not working correctly??

-   page.**waitForSelector**(myselector)  -> resolves race-issues
-  page.locator(myselector).**waitFor()**  ->  it doesn't and we need to add  a page.waitForTimeout(500) to avoid race conditions??

**Comment by AngelLopezBellmont** [comment]
@dgozman any idea about this?

**Comment by mxschmitt** [comment]
Sure its discouraged, but (what our guess it) the application behaves racy, then Playwright can also not fix it. 

If you are able to provide a smaller repro for us, maybe in form of just a single HTML site / the component which demonstrates this problem, that would help us a lot to find out if this is actually something Playwright can fix.

**Comment by AngelLopezBellmont** [comment]
@mxschmitt  hi sorry I dont get 100% your question. You mean the html of the problematic dialog? In the example above I create a whole tenant to show the problem https://dmitrygozman.fieldcode.com

But ok I can paste here the html dialog which is giving problems
 ```js
button the trigger the dialog:
<button class="MuiButtonBase-root MuiButton-root MuiButton-text jss147 jss1309 jss1012 jss151 jss1312 jss155 MuiButton-textSecondary MuiButton-textSizeLarge MuiButton-sizeLarge" tabindex="0" type="button">
  <span class="MuiButton-label">Cancel</span>
</button>

problematic dialog:
<div class="MuiDialog-container MuiDialog-scrollPaper" role="none presentation" tabindex="-1" style="opacity: 1; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;">
  <div class="MuiPaper-root MuiDialog-paper jss1324 jss1318 jss1322 MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded" role="dialog">
    <div class="jss1326">
      <div class="jss1329 jss1316"><span class="jss1336">Discard results</span></div>
      <div class="jss1332 jss1317"></div>
    </div>
    <div class="MuiDialogContent-root jss1330 jss1337">
      <div class="jss1320">Do you really want to discard the results?</div>
    </div>
    <div class="MuiDialogActions-root jss1334 jss1319 MuiDialogActions-spacing">
      <button
          class="MuiButtonBase-root MuiButton-root jss1335 MuiButton-text jss147 jss1338 jss149 jss1339 MuiButton-textPrimary"
          tabindex="0" type="button"><span class="MuiButton-label">Yes</span></button>
      <button class="MuiButtonBase-root MuiButton-root jss1335 MuiButton-text jss147 jss1345 jss151 jss1348 MuiButton-textSecondary" tabindex="0" type="button"><span class="MuiButton-label">No</span></button>
    </div>
  </div>
</div>
 ```
![2022_05_19_21_01_03_DevTools_dmitrygozman fieldcode com_](https://user-images.githubusercontent.com/12471931/169382229-87fac31c-71d6-4048-905d-00d7e101b4f0.png)


Please tell me if I need to paste any other info etc
Regards

**Comment by yury-s** [comment]
@AngelLopezBellmont the problem is that playwright clicks the button as soon as it becomes [actionable](https://playwright.dev/docs/next/actionability) which is enough in most cases but in your example the button is not immediately ready to handle clicks (perhaps click listener is not added / not properly setup yet), so you need to wait for a condition that is specific to your app to learn when the button becomes actionable (something like [waitForFunction](https://playwright.dev/docs/next/api/class-frame#frame-wait-for-function) may come in handy). With that said this is a real bug in the app as a human can also click the button sometimes with no effect and the best approach would be to fix the behavior so that the button doesn't show up or at least is disabled until it's ready to handled clicks.

Closing the issue as there is no generic solution on the playwright end.

**Comment by AngelLopezBellmont** [comment]
@yury-s yes you are right. I found the problem.  Aghh I can't belive it, it is 100% my fault 
I was checking the loading of the form before pressing click, but my check was not correct :(.  I was so focus about the button-cancel that I did not realize about that :( . I was checking the header and the buttons, which are automatically loaded, but the inside takes some miliseconds.  
And of course, as you said Yury, playwright was clicking button-cancel before the application could handle it (because the form was not yet loaded)
So everything can be solved, just adding a simple waitFor to any element inside the form.

 ```js
  //check form is open:
  await page.locator('text=Create ticket for project').waitFor({state:"visible"});
  await expect(page.locator('text=Create ticket for project')).toBeVisible;
  await page.locator('text=Last name*').waitFor({state:"visible"});     //with this new line everything works
 ```

Thanks again to all you for your help specially to @yury-s @mxschmitt
